### ISSUE TITLE
Matrix factorization bijectors

### ISSUE THREAD

**Issue by sethaxen** [TuringLang/Bijectors.jl#274]
There are a number of bijective matrix factorizations that it might make sense to include here, e.g.
- unique QR (usual QR but constraining the diagonal of R to be positive)
- unique SVD: usual SVD but constraining the first row of U to be positive
- polar decomposition (QP)
- LU factorization (if matrix is square with linearly independent columns)
- eigendecomposition (after fixing an eigenvalue sorting convention and eigenvector sign convention).
- symmetric eigendecomposition

For each of these, the logdetjacs are known and can be efficiently computed. The factorizations are especially common in random matrix theory. e.g. applying the unique QR polar decomposition to a matrix of IID std normal parameters gives a Q that is uniform on the Stiefel manifold; for the polar decomposition `sqrt(P)` is Wishart-distributed.

**Comment by sethaxen** [comment]
It probably makes sense for these bijectors to return a `Factorization` object but for their inverses to accept not only a `Factorization` but also an iterable of the factors. e.g. for unique QR, one needs to implement a new QR `Factorization` as well as an `AbstractQ` to enforce the sign convention. Since these might be of general interest, maybe it makes sense to make a UniqueFactorizations.jl package that only depends on LinearAlgebra, which could then be used in these bijectors.

**Comment by sethaxen** [comment]
Or does it make more sense to in an extension to a UniqueFactorizations.jl package, overload the Bijectors interface for functions like `UniqueFactorizations.qr`? Is there anything in Bijectors expecting all bijectors are a `Bijectors.Bijector` subtype?

**Comment by sethaxen** [comment]
@torfjelde @devmotion what do you think?

**Comment by torfjelde** [comment]
> Is there anything in Bijectors expecting all bijectors are a Bijectors.Bijector subtype?

Not really:) It really only lets you have a default implementation for stuff like `inverse` (which returns the lazy `Bijectors.Inverse` bijector).

You probably don't even need an extension (unless you want to hook it up to Distributions.jl using stuff like `bijector`); implementation of `ChangesOfVariables.with_logabsdet_jacobian` and `InverseFunctions.inverse` is sufficient to make it possible to use it with Bijectors.jl (but, as mentioned, for hooking into Distributions.jl, an extension might be the way to go).

**Comment by sethaxen** [comment]
I'm still undecided what to do about this, but here's an example of an efficient bijector for QR: https://gist.github.com/sethaxen/4d5a6d22e56794a90ece6224b8100f08. It's very similar to https://doi.org/10.1137/0717034 but also supports the following optional features:
- constraining `R` to be diagonal (for the most likely case that one only needs `Q`)
- constraining `det(Q)=1` for square `Q` (e.g. when one wants a rotation matrix)
- complex unitary `Q`

Note: This transform does still have singularities (analogous to the unit vector transform's singularity near 0), but there's a trick not yet implemented to increase the number of unconstrained parameters to increase the volume away from the singularities.
